simpleline/render/renderer.py

```diff
@@ -0,0 +1,382 @@
+        if renderer_stack:
+            self._screen_stack = renderer_stack
+        else:
+            self._screen_stack = ScreenStack()
+        self._register_handlers()
+
+        self.width = 80
+        self.redraw()
+
+    def _register_handlers(self):
+        self._event_loop.register_signal_handler(RenderScreenSignal, self._redraw_callback)
+        self._event_loop.register_signal_handler(InputScreenSignal, self._process_input_callback)
+
+    @property
+    def width(self):
+        return self._width
+
+    @width.setter
+    def width(self, width):
+        self._width = width
+        self._spacer = "\n".join(2 * [width * "="])
+
+    @property
+    def quit_screen(self):
+        return self._quit_screen
+
+    @quit_screen.setter
+    def quit_screen(self, quit_screen):
+        self._quit_screen = quit_screen
+
+    @property
+    def nothing_to_render(self):
+        """Is something for rendering in the renderer stack?
+
+        :return: True if the rendering stack is empty
+        :rtype: bool
+        """
+        return self._screen_stack.is_empty()
+
+    def schedule_screen(self, ui_screen, args=None):
+        """Add screen to the bottom of the stack.
+
+        This is mostly useful at the beginning to prepare the first screen hierarchy to display.
+
+        :param ui_screen: screen to show
+        :type ui_screen: UIScreen instance
+        :param args: optional argument, please see switch_screen for details
+        :type args: anything
+        """
+        screen = ScreenData(ui_screen, args)
+        self._screen_stack.add_first(screen)
+
+    def replace_screen(self, ui, args=None):
+        """Schedules a screen to replace the current one.
+
+        :param ui: screen to show
+        :type ui: instance of UIScreen
+        :param args: optional argument to pass to ui's refresh and setup methods
+                     (can be used to select what item should be displayed or so)
+        :type args: anything
+        """
+        try:
+            should_draw_immediately = self._screen_stack.pop().draw_immediately
+        except ScreenStackEmptyException:
+            raise ScreenStackEmptyException("Switch screen is not possible when there is no screen scheduled!")
+
+        # we have to keep the old_loop value so we stop
+        # dialog's mainloop if it ever uses switch_screen
+        screen = ScreenData(ui, args, should_draw_immediately)
+        self._screen_stack.append(screen)
+        self.redraw()
+
+    def switch_screen(self, ui, args=None):
+        """Schedules a screen to show, but keeps the current one in stack to
+        return to, when the new one is closed.
+
+        :param ui: screen to show
+        :type ui: UIScreen instance
+        :param args: optional argument
+        :type args: anything
+        """
+        screen = ScreenData(ui, args, False)
+        self._screen_stack.append(screen)
+        self.redraw()
+
+    def switch_screen_modal(self, ui, args=None):
```

Comment:
As noted earlier, but noting there as well for clarity: I think this should be push_screen_modal().